### PR TITLE
Fix JS error on dis/connect

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
The websocket handler was updated in https://github.com/mattermost/mattermost-plugin-github/pull/515.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/530
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/536

### CODE REVIEW

File: webapp/src/websocket/index.js [hunk 1]
```diff
@@ -19,9 +19,11 @@ export function handleConnect(store) {
         if (!msg.data) {
             return;
         }
+
         store.dispatch({
             type: ActionTypes.RECEIVED_CONNECTED,
-            data: {...msg.data, settings: {sidebar_buttons: Constants.SETTING_BUTTONS_TEAM, daily_reminder: true}},
+            data: {...msg.data, user_settings: {sidebar_buttons: Constants.SETTING_BUTTONS_TEAM, daily_reminder: true},
```
Comment (hanzei) [inline webapp/src/websocket/index.js:25]: This is the real fix

File: webapp/src/client/client.js [hunk 2]
```diff
@@ -79,10 +79,6 @@ export default class Client {
         return this.doGet(`${this.url}/pr?owner=${owner}&repo=${repo}&number=${prNumber}`);
     }
 
-    getSettings = async () => {
-        return this.doGet(`${this.url}/settings`);
-    }
```
Comment (hanzei) [inline webapp/src/client/client.js:84]: Cleanup

File: server/plugin/api.go [hunk 3]
```diff
@@ -484,6 +484,7 @@ func (p *Plugin) completeConnectUserToGitHub(c *Context, w http.ResponseWriter,
 			"github_client_id":    config.GitHubOAuthClientID,
 			"enterprise_base_url": config.EnterpriseBaseURL,
 			"organization":        config.GitHubOrg,
+			"plugin_settings":     config.ClientSafeSettings(),
```
Comment (hanzei) [inline server/plugin/api.go:487]: This ensure the redux state is updated and the client doesn't need to call `/connected` again

File: server/plugin/api.go [hunk 4]
```diff
@@ -555,20 +556,20 @@ func (p *Plugin) getConnected(c *Context, w http.ResponseWriter, r *http.Request
...
-		Organization      string             `json:"organization"`
-		UserSettings      *UserSettings      `json:"user_settings"`
-		PluginSettings    ClientSafeSettings `json:"plugin_settings"`
+		Connected         bool                   `json:"connected"`
+		GitHubUsername    string                 `json:"github_username"`
+		GitHubClientID    string                 `json:"github_client_id"`
+		EnterpriseBaseURL string                 `json:"enterprise_base_url,omitempty"`
+		Organization      string                 `json:"organization"`
+		UserSettings      *UserSettings          `json:"user_settings"`
+		PluginSettings    map[string]interface{} `json:"plugin_settings"`
```
Comment (hanzei) [inline server/plugin/api.go:565]: `map[string]interface{}` is used as `ClientSafeSettings` can't be send as a websocket event.

**Review by catalintomai — APPROVED** [review]

**Comment by hanzei** [comment]
@lufty-demansol Could you give this PR a review given that you worked on the issue before?

**Comment by lufty-demansol** [comment]
> @lufty-demansol Could you give this PR a review given that you worked on the issue before?

@hanzei Hey, sorry I was sick for several days. I'm actually haven't check this issue yet, still looking at gitlab issue. so go ahead.

**Comment by maisnamrajusingh** [comment]
> @lufty-demansol Could you give this PR a review given that you worked on the issue before?

@hanzei I think slight miscommunication here. I was the one looking into this.

File: webapp/src/websocket/index.js [hunk 5]
```diff
@@ -19,9 +19,10 @@ export function handleConnect(store) {
         if (!msg.data) {
             return;
         }
+
         store.dispatch({
             type: ActionTypes.RECEIVED_CONNECTED,
-            data: {...msg.data, settings: {sidebar_buttons: Constants.SETTING_BUTTONS_TEAM, daily_reminder: true}},
+            data: {...msg.data, user_settings: {sidebar_buttons: Constants.SETTING_BUTTONS_TEAM, daily_reminder: true}},
```
[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply webapp/src/websocket/index.js:25]: @hanzei for future purposes reference. What is the difference between `user_settings` and `settings` ?

File: server/plugin/api.go (hunk 4 above)
Comment (maisnamrajusingh) [reply server/plugin/api.go:565]: Another silly question as I am still new to this plugin. how is this prevented from sending out as a web socket event ?

**Review by maisnamrajusingh — APPROVED** [review]
I just have a few questions in general but otherwise this looks good.

File: webapp/src/websocket/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply webapp/src/websocket/index.js:25]: `settings` doesn't exist any longer. https://github.com/mattermost/mattermost-plugin-github/pull/515 renamed `settings` to `user_settings` in the websocket payload.

File: server/plugin/api.go (hunk 4 above)
Comment (hanzei) [reply server/plugin/api.go:565]: `PublishWebSocketEvent` accepts an `map[string]interface{}` as an argument. `gorp` is then used for encoding the request to the MM server. But `gorp` doesn't know how to handle arbitrary structs like `ClientSafeSettings`. Hence a `map[string]interface{}` needs to be used as `gorp` knows how to encode that.

**Comment by hanzei** [comment]
Heads up that I also included a fix for https://github.com/mattermost/mattermost-plugin-github/issues/536 in this PR given that I'm touching the code area anyway.

File: webapp/src/websocket/index.js [hunk 6]
```diff
@@ -34,12 +35,26 @@ export function handleDisconnect(store) {
...
 
+export function handleConfigurationUpdate(store) {
+    return (msg) => {
+        if (!msg.data) {
+            return;
+        }
+
+        store.dispatch({
+            type: ActionTypes.RECEIVED_CONFIGURATION,
+            data: {...msg.data},
```
Comment (mickmister) [inline webapp/src/websocket/index.js:53]: Seems weird to do this in this case 0/5

```suggestion
            data: msg.data,
```

File: webapp/src/websocket/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply webapp/src/websocket/index.js:25]: This code still seems fishy. We are hardcoding `user_settings` here and disregarding that data coming from the server.

```suggestion
            data: {
                    ...msg.data,
                    user_settings: {
                        sidebar_buttons: Constants.SETTING_BUTTONS_TEAM,
                        daily_reminder: true,
                        ...msg.data,
                    }
                },
```

This is the equivalent of "take all top level values from the websocket message besides `user_settings`. For the `user_settings` key, use the data from the server, and use these two defaults if the payload did not provide values for `sidebar_buttons` and/or `daily_reminder`."

File: server/plugin/api.go [hunk 7]
```diff
@@ -555,20 +556,20 @@ func (p *Plugin) getConnected(c *Context, w http.ResponseWriter, r *http.Request
...
-		Organization      string             `json:"organization"`
-		UserSettings      *UserSettings      `json:"user_settings"`
-		PluginSettings    ClientSafeSettings `json:"plugin_settings"`
+		Connected           bool                   `json:"connected"`
+		GitHubUsername      string                 `json:"github_username"`
+		GitHubClientID      string                 `json:"github_client_id"`
+		EnterpriseBaseURL   string                 `json:"enterprise_base_url,omitempty"`
+		Organization        string                 `json:"organization"`
+		UserSettings        *UserSettings          `json:"user_settings"`
+		ClientConfiguration map[string]interface{} `json:"configuration"`
```
Comment (mickmister) [inline server/plugin/api.go:565]: Why is gorp able to process `*UserSettings` but not `ClientSafeSettings`?

File: webapp/src/index.js [hunk 8]
```diff
@@ -42,6 +42,7 @@ class PluginClass {
 
         registry.registerWebSocketEventHandler(`custom_${pluginId}_connect`, handleConnect(store));
         registry.registerWebSocketEventHandler(`custom_${pluginId}_disconnect`, handleDisconnect(store));
+        registry.registerWebSocketEventHandler(`custom_${pluginId}_config_update`, handleConfigurationUpdate(store));
```
Comment (mickmister) [inline webapp/src/index.js:45]: I will never look at these registry functions the same ever again 🙂

File: server/plugin/configuration.go [hunk 9]
```diff
@@ -192,6 +199,8 @@ func (p *Plugin) OnConfigurationChange() error {
 
 	configuration.sanitize()
 
+	p.sendWebsocketEventIfneeded(p.getConfiguration(), configuration)
```
Comment (mickmister) [inline server/plugin/configuration.go:202]:
```suggestion
	p.sendWebsocketEventIfNeeded(p.getConfiguration(), configuration)
```

File: server/plugin/configuration.go [hunk 10]
```diff
@@ -216,6 +225,19 @@ func (p *Plugin) OnConfigurationChange() error {
 	return nil
 }
 
+func (p *Plugin) sendWebsocketEventIfneeded(oldConfig, newConfig *Configuration) {
+	// If the plugin just started, oldConfig is the zero value.
+	// Hence, an unnecessary websocket event is send.
```
Comment (mickmister) [inline server/plugin/configuration.go:230]:
```suggestion
	// Hence, an unnecessary websocket event is sent.
```

**Review by mickmister — CHANGES_REQUESTED** [review]
Just one change request, as I see we have the opportunity to fix a different bug visible in the diff

File: server/plugin/api.go (hunk 7 above)
Comment (hanzei) [reply server/plugin/api.go:565]: `ConnectedResponse`, which contains `UserSettings` is send a json response, not via websocket. `ClientConfiguration` could be a `struct` here. But not in https://github.com/mattermost/mattermost-plugin-github/blob/e5ff23f189aa15fe2d6bbdd0b325edfb79d91457/server/plugin/api.go#L479-L490 where it's send as a websocket payload.

File: webapp/src/websocket/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply webapp/src/websocket/index.js]: Good suggestion :+1:

**Comment by DHaussermann** [comment]
update-branch

**Review by mickmister — APPROVED** [review]
LGTM, nice work @hanzei!

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Tested that disconnect / reconnect  
- JS errors are resolved

LGTM!

**Pull request merged**